File: Data Structures & Algorithms/is-palindrome/submission-0.cpp
class Solution {
public:
    bool isPalindrome(string s) {
        int i = 0;
        int j = s.length() - 1;

        while (i < j) {
            // Skip non-alphanumeric characters from the left
            while (i < j && !std::isalnum(s[i])) {
                i++;
            }
            // Skip non-alphanumeric characters from the right
            while (i < j && !std::isalnum(s[j])) {
                j--;
            }
            
            // Compare the valid characters
            if (std::toupper(s[i]) != std::toupper(s[j])) {
                return false;
            }
           
            // Move pointers inward for the next check
            i++;
            j--;
        }
        return true;
    }
};

File: Data Structures & Algorithms/is-palindrome/submission-0.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        # convert string into lowercase
        # remove all whitespace in between characters
        # left and right pointer, and each time compare if they are the same char, if not then return false
        
        cleaned = "".join(char.lower() for char in s if char.isalnum())

        left = 0
        right = len(cleaned) - 1

        # use compare of left and right because for loop pointers will cross each other, you dont wanna cross
        while left < right:
            if cleaned[left] != cleaned[right]:
                return False

            left += 1
            right -= 1

        return True